language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -259,91 +258,335 @@ genModule pkgMap (Scope scope) curPkgId mod
+          , "  }),"
+          ]
+        | (n, d) <- serNestedDecoders
+        ]
+      , [ "});" ]
+      ]
+  | otherwise = assert (null serKeys) $ T.unlines $
+  -- If we have type parameters, the serializable definition is
+  -- a function and we generate extra properties on that function
+  -- for each nested decoder.
+      [ "export const " <> serName <> " = " <> tyArgs <> ":"
+      , "  damlTypes.Serializable<" <> serName <> tyParams <> "> => ({"
+      , "  decoder: () => " <> renderDecoder serDecoder <> ","
+      , "});"
+      ] <> concat
+      [ [ serName <.> n <> " = " <> tyArgs <> ":"
+        , "  damlTypes.Serializable<" <> serName <.> n <> tyParams <> "> => ({"
+        , "  decoder: () => " <> renderDecoder d <> ","
+        , "});"
+        ]
+      | (n, d) <- serNestedDecoders
+      ]
+  where tyParams = "<" <> T.intercalate ", " serParams <> ">"
+        tyArgs = tyParams <> "(" <> T.intercalate ", " (map (\name -> name <> ": damlTypes.Serializable<" <> name <> ">") serParams) <> ")"
+
+
+data Decoder
+    = DecoderOneOf T.Text [Decoder]
+    | DecoderObject [(T.Text, Decoder)]
+    | DecoderConstant DecoderConstant
+    | DecoderRef T.Text -- ^ Reference to an object with a .decoder() method
+    | DecoderLazy Decoder
+
+data DecoderConstant
+    = ConstantUndefined
+    | ConstantString T.Text -- ^ String literal
+    | ConstantRef T.Text -- ^ Variable reference
+
+renderDecoderConstant :: DecoderConstant -> T.Text
+renderDecoderConstant = \case
+    ConstantUndefined -> "undefined"
+    ConstantString s -> "'" <> s <> "'"
+    ConstantRef v -> v
+
+renderDecoder :: Decoder -> T.Text
+renderDecoder = \case
+    DecoderOneOf constr branches ->
+        "jtv.oneOf<" <> constr <> ">(" <>
+        T.concat (map (\b -> renderDecoder b <> ",") branches) <>
+        ")"
+    DecoderObject fields ->
+        "jtv.object({" <>
+        T.concat (map (\(name, d) -> name <> ":" <> renderDecoder d <> ",") fields) <>
+        "})"
+    DecoderConstant c -> "jtv.constant(" <> renderDecoderConstant c <> ")"
+    DecoderRef n -> n <> ".decoder()"
+    DecoderLazy d -> "jtv.lazy(() => " <> renderDecoder d <> ")"
+
+data TypeDef
+    = UnionDef T.Text [T.Text] [(T.Text, T.Text)]
+    | ObjectDef T.Text [T.Text] [(T.Text, T.Text)]
+    | EnumDef T.Text [T.Text] [T.Text]
+
+renderTypeDef :: TypeDef -> T.Text
+renderTypeDef = \case
+    UnionDef t args bs -> T.unlines $ concat
+        [ [ "export type " <> ty t args <> " =" ]
+        , [ "  |  { tag: '" <> k <> "'; value: " <> v <> " }" | (k, v) <- bs ]
+        ]
+    ObjectDef t args fs -> T.unlines $ concat
+        [ [ "export type " <> ty t args <> " = {" ]
+        , [ "  " <> k <> ": " <> v <> ";" | (k, v) <- fs ]
+        , [ "}" ]
+        ]
+    EnumDef t args fs -> T.unlines $ concat
+        [ [ "export type " <> ty t args <> " =" ]
+        , [ "  | '" <> f <> "'" | f <- fs ]
+        ]
+  where ty t args = t <> "<" <> T.intercalate ", " args <> ">"
```

Comment:
Shouldn't we treat the case when `args` is empty different?

Reply:
I realized that on my branch where I change it to emit declaration files but it doesn’t actually make a difference. Typescript seems to be perfectly happy with `<>`. (I’ll still fix it).